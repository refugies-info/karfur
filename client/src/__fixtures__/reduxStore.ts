import { RootState } from "services/rootReducer";
import { SearchResultsState } from "services/SearchResults/searchResults.reducer";
import { UserFavoritesState } from "services/UserFavoritesInLocale/UserFavoritesInLocale.reducer";
import { activeThemesMock } from "./activeThemes";

const initialMockUserState = {
  user: null,
  admin: false,
  traducteur: false,
  expertTrad: false,
  contributeur: false,
  membreStruct: false,
  userId: "",
  userFetched: false,
  rolesInStructure: [],
};
const initialMockLangueState = {
  langues: [],
  languei18nCode: "fr",
  showLanguageModal: false,
  showLangModal: false,
};

const initialMockTranslationState = {
  translation: {
    initialText: {},
    translatedText: {},
  },
  translations: [],
};

const initialUserFavoritesState: UserFavoritesState = {
  favorites: [],
  showFavoriteModal: false
}

const initialMockSearchReults: SearchResultsState = {
  results: {
    dispositifs: [],
    demarches: [],
    dispositifsSecondaryTheme: []
  },
  query: {
    search: "",
    departments: [],
    themes: [],
    needs: [],
    age: [],
    frenchLevel: [],
    language: [],
    sort: "date",
    type: "all",
  }
}

export const initialMockStore = {
  user: initialMockUserState,
  langue: initialMockLangueState,
  activeDispositifs: [],
  tts: { ttsActive: false, showAudioSpinner: false },
  userStructure: null,
  selectedDispositif: null,
  loadingStatus: {},
  translation: initialMockTranslationState,
  activeStructures: [],
  selectedStructure: null,
  allDispositifs: [],
  allStructures: [],
  users: [],
  userFavorites: initialUserFavoritesState,
  userContributions: [],
  needs: [],
  themes: {
    activeThemes: activeThemesMock,
    inactiveThemes: []
  },
  searchResults: initialMockSearchReults
};
