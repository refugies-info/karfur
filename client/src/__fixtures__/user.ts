import { GetUserInfoResponse, UserStatus } from "api-types";
import moment from "moment";

const date = moment("2019-05-01 12:00:00");
export const testUser: GetUserInfoResponse = {
  username: "test",
  _id: "testObjectId",
  structures: [],
  contributions: [],
  phone: "",
  roles: [],
  email: "",
  selectedLanguages: [],
  status: UserStatus.ACTIVE,
  traductionsFaites: []
};

export const testUserWithRoles: GetUserInfoResponse = {
  ...testUser,
  structures: ["testObjectId"],
  roles: [
    { nom: "Admin", _id: "testObjectId", nomPublic: "Admin" },
    { nom: "Trad", _id: "testObjectId", nomPublic: "Trad" },
    {
      nom: "ExpertTrad",
      _id: "testObjectId",
      nomPublic: "ExpertTrad"
    },
    {
      nom: "Contrib",
      _id: "testObjectId",
      nomPublic: "Contrib"
    }
  ]
};
