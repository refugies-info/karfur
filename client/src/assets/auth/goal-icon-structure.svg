<svg fill="none" height="48" viewBox="0 0 48 48" width="48" xmlns="http://www.w3.org/2000/svg"><path d="m9.70925 20.7096c.53855 1.6456 1.46905 2.2868 2.64475 3.7747.6127.7169.7514 1.7318.3531 2.5864-.8606 1.8502-.7195 4.4584-.9346 6.2947-.1705 1.4536-.6437 2.9408-1.4467 4.5466-.98559 1.9728-.2383 4.3862 2.0342 5.5082 4.2829 2.0529 4.3568 2.3006 5.0668 2.3004.6604 0 1.3085-.3741 1.6087-1.0358.0856-.1881.3302-.2476.4955-.0982.2054.1846 2.7363 2.4716 2.8638 2.5857 1.0642.962 3.2554.7189 4.3459.8181 1.3383.1217 2.4456-1.0285 2.2703-2.3624-.1222-.9314.3233-1.8394 1.1351-2.3135 1.1828-.691 2.184-1.5046 3.3085-1.0804l1.1607.4353c.0806.0303 2.7026 1.0129 2.7761 1.0401 1.4424.5411 2.907.0159 4.2576-.2818 1.6582-.3662 2.5709-1.7977 3.4827-2.9234.9512-1.1726.7248-2.8983-.496-3.7865l-.9764-.7097c-.1287-.0936-.2182-.2312-.2519-.3863l-.4287-1.9884c-.0602-.2778.0727-.5673.323-.7038 1.0935-.5965 1.4329-2.0127.7211-3.0409-.2127-.3521-.5415-.5892-.482-.9863l.1279-.8539c.1729-1.1505-.7613-2.163-1.9201-2.0875-.2766.0181-.4713-.2643-.3598-.5157l1.8163-4.0856c.1672-.3767.541-.6137.9545-.6034 1.2977.0334 2.422-.956 2.5247-2.279.2873-3.1485.1511-3.2782.4266-3.6367 1.0247-1.3395.3386-3.291-1.2982-3.696l-2.6939-.665c-.4445-.11-.4864-.3732-1.6728-1.62421-.6083-.67263-1.2506-.72344-2.0546-.86491-2.008-.35323-2.0383-.38096-4.1386-1.67755l-.0009-.00048c-1.2394-.70822-1.4224-1.19691-3.1646-3.09768-1.2899-1.45453-3.3061-2.358325-5.2617-2.358325h-1.0385c-.8899 0-1.6428.665455-1.7512 1.546465l-.3046 2.43308c-.0758.60835-.5553 1.0793-1.1651 1.14513-1.2832.13662-2.0892.21962-3.0258 1.18557-.0095.00979-.0407.04266-.0312.03287-.0822.08629-.139.14311-.191.24134-.2636.49732-.87.69997-1.3799.46164 0 0-3.0644-1.43583-3.1019-1.45347-1.7408-.81536-3.8063 1.00763-3.1333 2.84391l.7059 1.92852c.0889.243-.1376.4783-.3818.4046-4.80378-1.4519-4.6533-1.469-5.18698-1.3664l-2.88251.5492c-1.80554.3439-2.77575 2.3229-1.93993 3.9614.37747.7411.69377 1.3366 1.5994 1.7224l3.65142 1.5533c1.1429.4864 2.03612 1.4431 2.44065 2.6636zm37.37675-5.8381c-.038.0472-.0925.0966-.1599.1394.0592-.038.1132-.0849.1599-.1394zm-37.40419 5.7238c.00281.0183.0064.0365.01067.0545-.00311-.0127-.00708-.0314-.01067-.0545z" fill="#212121"/><g clip-rule="evenodd" fill-rule="evenodd"><path d="m34.2151 26.4739v-1.2911c-.2284-1.5723-.6547-4.0515-1.2819-5.8847-.7967-2.3286-3.2343-6.3241-4.1582-7.7992-.8477-1.1902-1.3463-2.64636-1.3463-4.21892 0-4.02089 3.2596-7.28046828 7.2805-7.28046828s7.2804 3.25957828 7.2804 7.28046828c0 1.57413-.4995 3.03162-1.3488 4.22242-.9534 1.5089-3.3975 5.484-4.1885 7.7957-.6272 1.8332-1.0535 4.3124-1.2818 5.8847v1.2911z" fill="#c1c1fb"/><path d="m34.7095 10.5888c1.4621 0 2.6474-1.18527 2.6474-2.64741s-1.1853-2.64744-2.6474-2.64744c-1.4622 0-2.6475 1.1853-2.6475 2.64744s1.1853 2.64741 2.6475 2.64741z" fill="#fff"/><path d="m16.2151 37.617v-1.2912c-.2284-1.5722-.6547-4.0514-1.2819-5.8846-.7967-2.3286-3.2343-6.3241-4.1582-7.7993-.84771-1.1902-1.34629-2.6463-1.34629-4.2189 0-4.0208 3.25959-7.2804 7.28049-7.2804s7.2804 3.2596 7.2804 7.2804c0 1.5742-.4995 3.0316-1.3488 4.2225-.9534 1.5089-3.3975 5.484-4.1885 7.7957-.6272 1.8332-1.0535 4.3124-1.2818 5.8846v1.2912z" fill="#c1c1fb"/><path d="m16.7095 21.7319c1.4621 0 2.6474-1.1853 2.6474-2.6474 0-1.4622-1.1853-2.6475-2.6474-2.6475-1.4622 0-2.6475 1.1853-2.6475 2.6475 0 1.4621 1.1853 2.6474 2.6475 2.6474z" fill="#fff"/></g></svg>