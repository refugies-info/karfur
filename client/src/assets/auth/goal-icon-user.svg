<svg fill="none" height="56" viewBox="0 0 27 56" width="27" xmlns="http://www.w3.org/2000/svg"><path d="m5.51172 25.9609 4.60998 15.8595 2.1718 11.9049h2.1718l-2.1718-27.7644z" fill="#ac5419"/><path d="m3.44193 25.9609c.32782 8.2371.18439 12.7245.12293 13.4826-.08196.7582-.47125 5.7578-3.257731 14.5277h2.253771c3.68799-8.4215 4.9583-13.4417 5.38857-14.5277.43026-1.1064 1.65959-5.5938 3.48313-13.4826z" fill="#ac5419"/><path d="m5.42969 25.9609c1.18835 5.4505 3.25772 14.1588 6.20811 26.1252h3.1757c.2254-12.2737-.758-20.3674-2.4996-26.1252z" fill="#212121"/><path d="m3.44202 25.9609c.30733 8.2371-.49174 16.3513-3.032352 26.2686h3.380652c3.68799-8.2986 6.33108-16.3718 8.13408-26.2686z" fill="#5e5e5e"/><path d="m0 55.6104.122953-2.0081c.737598.2049 1.639087.123 2.745487-.2459 1.18835.8402 2.66355 1.4139 4.46657 1.7212.14342.0205.22538.1435.20489.2869v.0205l-.16392.6147h-4.48704-2.622575z" fill="#212121"/><path d="m11.8018 55.6104.1229-2.0081c.7376.2049 1.6391.123 2.7455-.2459 1.1883.8402 2.6635 1.4139 4.4665 1.7212.1435.0205.2254.1435.2049.2869v.0205l-.1639.6147h-4.487-2.6226z" fill="#212121"/><path d="m8.46194 6.45418c-.57369-.67618-.96298-1.51628-.90151-2.43834.1844-2.68423 3.93387-2.09001 4.65097-.73765.7376 1.35235.6352 4.79472-.3073 5.04061-.3688.10245-1.1679-.14344-1.98746-.6352l.51226 3.6063h-2.9914z" fill="#ac5419"/><path d="m13.5635 1.39285c-.0615-.22539-.2254-.409804-.4508-.471275-.0614-.02049-.1024-.020491-.1639-.020491-.0205 0-.0205 0-.041 0 0-.020491 0-.040977 0-.040977 0-.040981-.0204-.081963-.0409-.122943-.041-.122942-.123-.245885-.2254-.327846-.1844-.184413-.4303-.286866-.6761-.3278463-.1435-.0204903-.2664.0204943-.4098.0614743-.0615.020491-.123.06147-.2049.08196-.0205 0-.082.061469-.1025.061469-.0204 0-.0614-.061469-.0819-.061469-.2049-.1639221-.4713-.22539428-.7376-.22539428-.3073 0-.59418.02049028-.84005.20490328-.06146.04098-.10243.081962-.14341.122943-.02049.02049-.04098.040979-.08196.08196-.02049.02049-.0205.040983-.04099.040983-.02049.02049-.02048.040975-.04097.020485-.12293-.081961-.26635-.10245-.40978-.08196-.08195.02049-.1639.040985-.24585.061475-.02049.02049-.08198.061468-.12295.061468-.02049 0-.08195-.040978-.12293-.061468-.16391-.061471-.32783-.061474-.49174-.040983-.16391.02049-.3278.10245-.47123.184411-.06146.040981-.12293.102455-.1844.163926-.02049.020491-.04099.06147-.06148.08196-.02049.020491-.04097.040977-.04097.040977 0-.02049-.18439-.08196-.20488-.08196-.10245-.02049-.18441-.020489-.28685.020492-.16391.061471-.32781.184416-.45075.307356-.06146.06147-.12294.14343-.16392.22539-.02049.04098-.04096.08197-.06145.12295 0 .02049-.02051.04097-.02051.06146-.02049.02049-.02047.0205-.06145.04099-.14342.04098-.26636.14343-.36881.24588-.10244.12294-.16391.26637-.1844.4098v.04099h-.04097c-.04098 0-.08195.02049-.10244.02049-.06147.02049-.14343.04098-.20489.06147-.24587.12294-.40978.34833-.49174.59421-.08195.26638-.08196.57373.02049.84011.02048.04098.04098.08196.06147.12294.02049.02049.02049.02049 0 .04098s-.02048.06147-.04097.08196c-.04097.08196-.06147.18441-.08196.28687-.02049.16392-.04097.32784-.02048.47127.02049.08196.04096.14343.08194.2049.02048.04098.04098.06147.06147.10246.02049.02049.02048.02049.04097.04098l.02051.02049c.02049.02049.02048 0 .02048.04098-.02049.04098-.04099.08196-.06147.12294-.02049.04098-.04097.08196-.04097.14343-.02049.10246-.06148.18442-.06148.28687-.02049.18441-.02048.38931.06148.55323.04098.08196.08194.14344.16389.18442.04098.02049.06147.04098.10245.06147.02049 0 .04099.02049.04099.02049v.08196c-.02049.18441-.04098.38931.02048.57373.04098.16392.16391.32784.30734.4098.28684.18441.67612.18442.96296.02049.08196-.04098.14344-.08195.2254-.14343.16391.2254.51221.22539.73759.14343.30733-.08196.5327-.28686.67612-.55323.1844.22539.59417.16392.71711-.10245.04097-.06147.06147-.12294.06147-.20491 0-.04098 0-.08196 0-.10245 0-.04098.04099-.08196.06148-.14343.04098-.08196.06145-.18441.06145-.26637 0-.04098.00001-.10246-.02048-.14344 0-.04098-.00001-.08196.02048-.12294-.08195.02049-.1639 0-.24586-.02049-.06146-.02049-.14342-.08196-.1844-.12294-.12293-.12294-.18439-.28687-.24586-.45079-.10244-.34833-.06148-.77863.28683-.96305.16391-.08196.65566-.08195.86054-.04097.20489.04098.43026.49176.45075.55323 0 .02049-.00001.06148.02048.08197.06147.04098.14343-.0205.18441-.06148.06146-.06147.12294-.14343.16392-.2049.08195-.16392.16395-.34834.22535-.53275.041-.14343.0615-.28686.1229-.43029.0615-.14343.1435-.26638.2664-.34834.1434-.08196.3073-.12294.4508-.18441.1434-.06147.3073-.12294.4097-.22539.0205-.02049.041-.04098.082-.08196.0205-.02049.041-.06147.0615-.08196.0409-.02049.1024.04097.1229.06147.1639.10245.3073.24588.4098.4098.1024.16392.2048.30735.2663.49177.0615.12294.082.30735.1844.4098.041.04098.3688.08197.4713.04099.1024-.04098.2253-.14344.2458-.2254.041-.06147.041-.14343.0205-.22539 0-.04098-.0205-.08197-.041-.12295 0-.04098-.0204-.08196-.0409-.10245-.041-.06147-.1025-.10244-.1639-.12293.0819-.06148.1434-.14344.1844-.24589.0409-.10245.0819-.2049.0819-.30735 0-.08197 0-.16393-.041-.24589s-.1434-.12294-.2049-.16392c.123-.08196.2049-.2049.2869-.30736.1639-.28686.2663-.57373.1844-.86059z" fill="#212121"/><path d="m12.375 12.888 4.3641-2.3974c.7171-.86059 1.4343-1.53677 2.1514-2.00805.2049-.08196.5941-.14343.2458.47128s-.6966 1.29087-.5122 1.49577c.1844.1844.6352-.0819.84.2459.1435.2049-.6146.5532-2.2947 1.004l-3.1758 2.9096z" fill="#ac5419"/><path d="m9.99836 12.56c1.53664-.0204 3.56504-.3483 5.73684-1.7211l.8605 1.9056c-1.4956 1.4958-3.606 2.7047-5.45 2.6637-1.45468-.0615-2.19227-1.8031-1.14734-2.8482z" fill="#adadf9"/><path d="m26.6967 6.1669c-1.967-.45079-3.9134-1.00402-5.8598-1.51628-.2459-.06147-1.2089-.45079-1.3933-.36883 0 0-.1639 1.02452-.2049 1.31138-.1024.98353-.2458 1.94658-.3893 2.93011-.1024.67618-.1843 1.24991-.2458 1.55722-.0205.0615-.0205.123-.041.164-.0205.1024 0 .1639.041.2254.1844.1844.6351-.082.84.2458.1025.164-.3278.4098-1.3317.7377-.5328 1.2704-2.3972 3.9751-2.5816 4.2005 1.4957.6557 2.9913 1.3114 4.5075 1.9671.3688.1639 2.5816 1.2294 2.9709 1.1064.3893-.1024.9835-1.3318 1.2908-1.9875.7581-1.5983 1.3522-3.2785 1.762-4.9996.1025-.4918 1.0245-5.47095.6352-5.5734z" fill="#6a6af4"/><path d="m19.792 17.4572 2.0694-1.1475c.2868-1.1065.6351-1.6597 1.0449-1.6802.3073-.1434-.2254 1.2499.3688 1.127.5942-.123 2.0284-1.1885 2.2743-.9836.3687.3279.0409 1.3114-.3484 1.8032-.7171.9221-1.3113 1.2294-2.7865 1.6392-.7785.2049-1.5981.6352-2.4996 1.2909z" fill="#ac5419"/><path d="m12.8877 16.9859c1.6186.4098 4.3436-.0819 8.1545-1.4343.9425 1.9671 1.3933 3.3604 1.3113 4.18-3.2782 1.7007-6.0852 2.4794-8.4004 2.5203.0819 1.3729-.1025 2.6843-.6557 3.8727-1.1473 2.4588-7.90863.1844-10.18289.6352-1.41372-4.6103 1.33178-6.4954 1.33178-9.446 0-3.2375 2.60207-8.21663 2.99135-8.21663h3.48306c.2049 2.47933 1.1884 5.20453 1.967 7.88873z" fill="#c1c1fb"/></svg>