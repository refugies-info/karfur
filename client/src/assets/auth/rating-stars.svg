<svg fill="none" height="24" viewBox="0 0 137 24" width="137" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink"><linearGradient id="a" gradientUnits="userSpaceOnUse" x1="130.206" x2="130.777" y1="11.9001" y2="11.9001"><stop offset="0" stop-color="#6a6af4"/><stop offset="1" stop-color="#6a6af4" stop-opacity=".4"/></linearGradient><g clip-rule="evenodd" fill-rule="evenodd"><g fill="#6a6af4"><path d="m12.4999 18.26-7.05299 3.948 1.575-7.928-5.935-5.488 8.02701-.952 3.38598-7.34 3.386 7.34 8.027.952-5.935 5.488 1.575 7.928z"/><path d="m40.4999 18.26-7.053 3.948 1.575-7.928-5.935-5.488 8.027-.952 3.386-7.34 3.386 7.34 8.027.952-5.935 5.488 1.575 7.928z"/><path d="m68.4999 18.26-7.053 3.948 1.575-7.928-5.935-5.488 8.027-.952 3.386-7.34 3.386 7.34 8.027.952-5.935 5.488 1.575 7.928z"/><path d="m96.4999 18.26-7.053 3.948 1.575-7.928-5.935-5.488 8.027-.952 3.386-7.34 3.386 7.34 8.0271.952-5.935 5.488 1.575 7.928z"/></g><path d="m124.5 18.26-7.053 3.948 1.575-7.928-5.935-5.488 8.027-.952 3.386-7.34 3.386 7.34 8.027.952-5.935 5.488 1.575 7.928z" fill="url(#a)"/></g></svg>