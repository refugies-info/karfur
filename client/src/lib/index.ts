export const jsUcfirst = (string: string) =>
  string &&
  string.length > 1 &&
  string.charAt(0).toUpperCase() + string.slice(1, string.length);

export const jsLcfirst = (string: string) =>
  string &&
  string.length > 1 &&
  string.charAt(0).toLowerCase() + string.slice(1, string.length);

export const limitNbCaracters = (string: string, nbCaractersMax: number) =>
  string.substring(0, Math.min(string.length, nbCaractersMax)) +
  (string.length > nbCaractersMax ? "..." : "");

export const removeAccents = (str = "") => {
  var accents =
    "ÀÁÂÃÄÅàáâãäåÒÓÔÕÕÖØòóôõöøÈÉÊËèéêëðÇçÐÌÍÎÏìíîïÙÚÛÜùúûüÑñŠšŸÿýŽž";
  var accentsOut =
    "AAAAAAaaaaaaOOOOOOOooooooEEEEeeeeeCcDIIIIiiiiUUUUuuuuNnSsYyyZz";
  const chars = str.split("");
  let i, x;
  for (i = 0; i < chars.length; i++) {
    if ((x = accents.indexOf(chars[i])) !== -1) {
      chars[i] = accentsOut[x];
    }
  }
  return chars.join("");
};

// rules for password check :
// nbNumUpper = nbNumeric + nbUpper
// score 0 (cannot set password): length < 5
// score 0.5 : 7>length>=5 + nbNumUpper = 1 or 2 or0
// score 1.5 : 7>length >= 5 + nbNumUpper>=3 OR length>=7 + nbNumUpper<3
// score 2.5 : length >= 7 + nbNumUpper>=3
// score 3.5 : length >= 7 + nbNumUpper>=3 + at least 1 special caractere

// make sure to use same function in backend
export const computePasswordStrengthScore = (
  password: string
): { score: number } => {
  if (password.length < 5) {
    return { score: 0 };
  }

  const numerics = password.match(/\d+/g)?.join("");
  const nbNumeric = numerics ? numerics.length : 0;

  const upperCase = password.match(/[A-Z]/g)?.join("");
  const nbUpperCase = upperCase ? upperCase.length : 0;

  if (password.length < 7) {
    if (nbUpperCase + nbNumeric < 3) {
      return { score: 1 };
    }
    return { score: 2 };
  }

  if (nbNumeric + nbUpperCase < 3) {
    return { score: 2 };
  }
  const specialCar = password.match(/[!@#$&*]/g)?.join("");
  const nbSpecialCar = specialCar ? specialCar.length : 0;
  if (nbSpecialCar > 0) {
    return { score: 4 };
  }

  return { score: 3 };
};
