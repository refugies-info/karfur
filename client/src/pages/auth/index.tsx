import React, { ReactElement, useCallback, useState } from "react";
import { useAsyncFn } from "react-use";
import { useRouter } from "next/router";
import Image from "next/image";
import { Button } from "@codegouvfr/react-dsfr/Button";
import { Input } from "@codegouvfr/react-dsfr/Input";
import { getPath } from "routes";
import { logger } from "logger";
import { useAuthRedirect, useLogin, useRegisterFlow } from "hooks";
import { googleProvider } from "utils/googleSignIn";
import API from "utils/API";
import { defaultStaticProps } from "lib/getDefaultStaticProps";
import { cls } from "lib/classname";
import { isValidEmail } from "lib/validateFields";
import { getRegisterInfos } from "lib/loginRedirect";
import { Event } from "lib/tracking";
import SEO from "components/Seo";
import ErrorMessage from "components/UI/ErrorMessage";
import Layout from "components/Pages/auth/Layout";
import Loader from "components/Pages/auth/Loader";
import GoogleIcon from "assets/auth/providers/google-icon.svg";
import MicrosoftIcon from "assets/auth/providers/microsoft-icon.svg";
import OutlookIcon from "assets/auth/providers/outlook-icon.svg";
// import DataInclusionIcon from "assets/auth/providers/data-inclusion-icon.svg";
import styles from "scss/components/auth.module.scss";

const AuthEmail = () => {
  useAuthRedirect(false);
  const router = useRouter();
  const { logUser, handleError } = useLogin();
  const { start } = useRegisterFlow(null);
  const [email, setEmail] = useState("");
  const [formError, setFormError] = useState("");
  const [error, setError] = useState("");
  const [isLoading, setIsLoading] = useState(false);

  const [{ loading }, submit] = useAsyncFn(
    async (e: any) => {
      e.preventDefault();
      if (!isValidEmail(email)) {
        setFormError("Veuillez entrer une adresse email valide");
        return;
      }
      try {
        Event("AUTH", "password login", "start");
        const res = await API.checkUserExists(email);
        router.push(getPath("/auth/connexion", "fr", `?email=${email}${res.verificationCode ? "&2fa=true" : ""}`));
      } catch (e) {
        router.push(getPath("/auth/inscription", "fr", `?email=${email}`));
      }
    },
    [router, email],
  );

  const loginGoogle = useCallback(() => {
    if (!googleProvider) {
      logger.error("[loginGoogle] Wrong Google provider configuration");
      return;
    }
    Event("AUTH", "google login", "start");
    googleProvider.useGoogleLogin({
      flow: "auth-code",
      onSuccess: ({ code }) => {
        setIsLoading(true);
        const registerInfos = getRegisterInfos();
        API.login({
          authGoogle: {
            authCode: code,
          },
          role: registerInfos?.role, // set role in case new account
        })
          .then((res) => {
            Event("AUTH", "google login", "success");
            if (res.userCreated) start(res.token, registerInfos?.role);
            else logUser(res.token);
          })
          .catch((e) => {
            setIsLoading(false);
            const error = handleError(
              e.response?.data?.code,
              e.response?.data?.data?.email || "",
              e.response?.data?.data,
            );
            if (error) setError(error);
          });
      },
      onError: (err) => {
        setIsLoading(false);
        logger.error("[loginGoogle] Failed to login with google", err);
        setError("Erreur, vous n'êtes pas authentifié avec votre compte Google, veuillez réessayer.");
      },
    })();
  }, [logUser, start, handleError]);

  const loginMicrosoft = useCallback(async () => {
    try {
      setIsLoading(true);
      Event("AUTH", "microsoft login", "start");
      await API.login({
        authMicrosoft: {
          authCode: null, // send a null code to get the auth url
        },
      });
    } catch (e: any) {
      const errorCode = e.response?.data?.code;
      if (errorCode === "SSO_URL") {
        // and redirect if url is provided
        if (e.response.data.data.url) {
          window.location.href = e.response.data.data.url;
        } else {
          setIsLoading(false);
          setError("Erreur, vous n'êtes pas authentifié avec votre compte Microsoft, veuillez réessayer.");
        }
      }
    }
  }, []);
  // const loginInclusionConnect = useCallback(() => {}, []);

  return (
    <div className={cls(styles.container, styles.half)}>
      <SEO
        title="Bienvenue"
        description="Votre compte Réfugiés.info vous permet d’avoir une expérience personnalisée."
      />

      {isLoading ? (
        <Loader text="Connexion en cours..." />
      ) : (
        <>
          <Button priority="tertiary" size="small" iconId="fr-icon-arrow-left-line" onClick={() => router.back()}>
            Retour
          </Button>
          <div className={styles.title}>
            <h1>Bienvenue&nbsp;!</h1>
            <p className={styles.subtitle}>
              Votre compte Réfugiés.info vous permet d’avoir une expérience personnalisée.
            </p>
          </div>

          <form onSubmit={submit}>
            <Input
              label="Adresse mail"
              state={!formError ? "default" : "error"}
              stateRelatedMessage={formError}
              nativeInputProps={{
                autoFocus: true,
                type: "email",
                value: email,
                onChange: (e: any) => setEmail(e.target.value),
              }}
            />

            <Button
              iconId="fr-icon-arrow-right-line"
              iconPosition="right"
              className={styles.button}
              nativeButtonProps={{ type: "submit" }}
              disabled={loading}
            >
              Continuer
            </Button>
          </form>

          <div className={cls(styles.separator, styles.mx)}>
            <span>ou connectez-vous avec</span>
          </div>

          <Button onClick={loginGoogle} className={cls(styles.button, "mb-4")} priority="tertiary">
            <Image src={GoogleIcon} width={24} height={24} alt="" className="me-2" />
            Google
          </Button>
          <Button onClick={loginMicrosoft} className={cls(styles.button, "mb-4")} priority="tertiary">
            <Image src={MicrosoftIcon} width={24} height={24} alt="" className="me-2" />
            <Image src={OutlookIcon} width={24} height={24} alt="" className="me-2" />
            Microsoft ou Outlook
          </Button>
          {/* <Button onClick={loginInclusionConnect} className={cls(styles.button, "mb-4")} priority="tertiary">
          <Image src={DataInclusionIcon} width={24} height={24} alt="" className="me-2" />
          Inclusion Connect
        </Button> */}

          <ErrorMessage error={error} />
        </>
      )}
    </div>
  );
};

export const getStaticProps = defaultStaticProps;
export default AuthEmail;

// override default layout and options
AuthEmail.getLayout = (page: ReactElement) => <Layout loginHelp>{page}</Layout>;
