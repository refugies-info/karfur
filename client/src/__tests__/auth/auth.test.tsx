import "jest-styled-components";
import { initialMockStore } from "__fixtures__/reduxStore";
import { wrapWithProvidersAndRenderForTesting } from "../../../jest/lib/wrapWithProvidersAndRender";
import auth from "../../pages/auth";
import { setupGoogleMock } from "../../__mocks__/react-google-autocomplete";

jest.mock("next/router", () => require("next-router-mock"));

describe("auth", () => {
  beforeEach(() => {
    jest.clearAllMocks();
    setupGoogleMock();
  });

  it("renders auth", () => {
    window.scrollTo = jest.fn();
    const component = wrapWithProvidersAndRenderForTesting({
      Component: auth,
      reduxState: {
        ...initialMockStore,
      },
    });
    expect(component).toMatchSnapshot();
  });
});
