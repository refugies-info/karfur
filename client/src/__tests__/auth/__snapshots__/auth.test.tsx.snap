// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`auth renders auth 1`] = `
<div
  className="container half"
>
  <button
    className="fr-btn fr-btn--tertiary fr-btn--sm fr-icon-arrow-left-line fr-btn--icon-left"
    id="fr-button-:r0:"
    onClick={[Function]}
    style={{}}
  >
    Retour
  </button>
  <div
    className="title"
  >
    <h1>
      Bienvenue !
    </h1>
    <p
      className="subtitle"
    >
      Votre compte Réfugiés.info vous permet d’avoir une expérience personnalisée.
    </p>
  </div>
  <form
    onSubmit={[Function]}
  >
    <div
      className="fr-input-group"
    >
      <label
        className="fr-label"
        htmlFor="input-:r1:"
      >
        Adresse mail
      </label>
      <input
        aria-describedby="input-:r1:-desc-error"
        autoFocus={true}
        className="fr-input"
        id="input-:r1:"
        onChange={[Function]}
        type="email"
        value=""
      />
    </div>
    <button
      className="fr-btn fr-icon-arrow-right-line fr-btn--icon-right button"
      disabled={false}
      id="fr-button-:r2:"
      style={{}}
      type="submit"
    >
      Continuer
    </button>
  </form>
  <div
    className="separator mx"
  >
    <span>
      ou connectez-vous avec
    </span>
  </div>
  <button
    className="fr-btn fr-btn--tertiary button mb-4"
    id="fr-button-:r3:"
    onClick={[Function]}
    style={{}}
  >
    <img
      alt=""
      className="me-2"
      data-nimg="1"
      decoding="async"
      height={24}
      loading="lazy"
      onError={[Function]}
      onLoad={[Function]}
      src="/_next/image?url=%2Fimg.jpg&w=48&q=75"
      srcSet="/_next/image?url=%2Fimg.jpg&w=32&q=75 1x, /_next/image?url=%2Fimg.jpg&w=48&q=75 2x"
      style={
        {
          "color": "transparent",
        }
      }
      width={24}
    />
    Google
  </button>
  <button
    className="fr-btn fr-btn--tertiary button mb-4"
    id="fr-button-:r4:"
    onClick={[Function]}
    style={{}}
  >
    <img
      alt=""
      className="me-2"
      data-nimg="1"
      decoding="async"
      height={24}
      loading="lazy"
      onError={[Function]}
      onLoad={[Function]}
      src="/_next/image?url=%2Fimg.jpg&w=48&q=75"
      srcSet="/_next/image?url=%2Fimg.jpg&w=32&q=75 1x, /_next/image?url=%2Fimg.jpg&w=48&q=75 2x"
      style={
        {
          "color": "transparent",
        }
      }
      width={24}
    />
    <img
      alt=""
      className="me-2"
      data-nimg="1"
      decoding="async"
      height={24}
      loading="lazy"
      onError={[Function]}
      onLoad={[Function]}
      src="/_next/image?url=%2Fimg.jpg&w=48&q=75"
      srcSet="/_next/image?url=%2Fimg.jpg&w=32&q=75 1x, /_next/image?url=%2Fimg.jpg&w=48&q=75 2x"
      style={
        {
          "color": "transparent",
        }
      }
      width={24}
    />
    Microsoft ou Outlook
  </button>
</div>
`;
